File: Linked-Lists/reverseLinkedList_Leetcode.cpp
/*
    Problem: https://leetcode.com/problems/reverse-linked-list/
*/

#include <iostream>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        
        ListNode* prev = NULL;
        ListNode* future = NULL;
        ListNode* curr = head;
        
        while(curr != NULL){
            future = curr->next;    // preserve curr.next 
            curr->next = prev;      // reroute curr.next to prev (reversing)
            prev = curr;            // move prev up to curr's position
            curr = future;          // move curr up to future's position
        }
        
       
        return prev;  // prev is now new 'head' of reversed-LL
    }
};

File: Linked-Lists/reverseLinkedList_leetcode.java
// problem: https://leetcode.com/problems/reverse-linked-list/
// TIME: O(n)
// SPACE: O(1)
// - Most optimal solution for reversing a linked list implemented in Java!

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        
        ListNode curr = head;
        ListNode prev = null;
        ListNode future = null;
        while(curr != null){
            future = curr.next;
            curr.next = prev;
            prev = curr;
            curr = future;
        }
        
        return prev;
    }
}